domain/application/state/application_test.go

```diff
@@ -3028,6 +3030,266 @@ func (s *applicationStateSuite) TestHashConfigAndSettings(c *gc.C) {
+		_, err := tx.ExecContext(ctx, addConstraintStmt, "constraint-uuid", "amd64", 2, true, "image-id")
+		if err != nil {
+			return err
+		}
+		addAppConstraintStmt := `INSERT INTO application_constraint (application_uuid, constraint_uuid) VALUES (?, ?)`
+		_, err = tx.ExecContext(ctx, addAppConstraintStmt, id, "constraint-uuid")
+		return err
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	cons, err := s.state.GetApplicationConstraints(context.Background(), id)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(cons, jc.DeepEquals, constraints.Value{
+		Arch:             ptr("amd64"),
+		CpuCores:         ptr(uint64(2)),
+		AllocatePublicIP: ptr(true),
+		ImageID:          ptr("image-id"),
+	})
+}
+
+func (s *applicationStateSuite) TestConstraintEmpty(c *gc.C) {
+	id := s.createApplication(c, "foo", life.Alive)
+
+	cons, err := s.state.GetApplicationConstraints(context.Background(), id)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(cons, jc.DeepEquals, constraints.Value{})
+}
+
+func (s *applicationStateSuite) TestSetConstraintFull(c *gc.C) {
+	id := s.createApplication(c, "foo", life.Alive)
+
+	cons := constraints.Value{
+		Arch:             ptr("amd64"),
+		CpuCores:         ptr(uint64(2)),
+		CpuPower:         ptr(uint64(42)),
+		Mem:              ptr(uint64(8)),
+		RootDisk:         ptr(uint64(256)),
+		RootDiskSource:   ptr("root-disk-source"),
+		InstanceRole:     ptr("instance-role"),
+		InstanceType:     ptr("instance-type"),
+		Container:        ptr(instance.LXD),
+		VirtType:         ptr("virt-type"),
+		AllocatePublicIP: ptr(true),
+		ImageID:          ptr("image-id"),
+		Spaces:           ptr([]string{"space0", "space1"}),
+		Tags:             ptr([]string{"tag0", "tag1"}),
+		Zones:            ptr([]string{"zone0", "zone1"}),
+	}
+
+	err := s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		insertSpace0Stmt := `INSERT INTO space (uuid, name) VALUES (?, ?)`
+		_, err := tx.ExecContext(ctx, insertSpace0Stmt, "space0-uuid", "space0")
+		if err != nil {
+			return err
+		}
+		insertSpace1Stmt := `INSERT INTO space (uuid, name) VALUES (?, ?)`
+		_, err = tx.ExecContext(ctx, insertSpace1Stmt, "space1-uuid", "space1")
+		return err
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.state.SetApplicationConstraints(context.Background(), id, cons)
+	c.Assert(err, jc.ErrorIsNil)
+
+	var (
+		applicationUUID  string
+		constraintUUID   string
+		constraintSpaces []string
+		constraintTags   []string
+		constraintZones  []string
+	)
+	err = s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		err := tx.QueryRowContext(ctx, "SELECT application_uuid, constraint_uuid FROM application_constraint WHERE application_uuid=?", id).Scan(&applicationUUID, &constraintUUID)
+		if err != nil {
+			return err
+		}
+		c.Check(constraintUUID, gc.Not(gc.Equals), "")
+		c.Check(applicationUUID, gc.Equals, id.String())
```

Comment:
No checks inside a transaction!!

Reply:
Yeah! you were quicker than static analysis..
Fixed